### PR TITLE
Keymapping unit tests

### PR DESCRIPTION
**Pull request by dlyr**
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
unit tests

* **What is the current behavior?** (You can also link to an open issue here)
keymapping manager is not tested at all

* **What is the new behavior (if this is a feature change)?**
basic loading is tested

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
nope.

* **Other information**:
todo (now of for the futur depending on time)
- [x] add integration tests for file saving
- [x] add  keymappingmanageable test with some dummy classes (also tests the observable behavior of KeyMappingManager).

### CODE REVIEW

**Comment by dlyr** [comment]
I saw that KeyMappingManager::addAction directly save the configuration file. 
I do not get the point why this is a good behavior.
KeyMappingManager could have a "modified" state that application could watch, or that can trigger an "ask to save before reload or quit".
In its current form it's hard to unittest ... and the behavior seems misleading.
Any comments welcome.

File: src/Gui/Utils/KeyMappingManager.cpp [hunk 1]
```diff
@@ -227,16 +227,17 @@ void KeyMappingManager::bindKeyToAction( Ra::Core::Utils::Index contextIndex,
...
-void KeyMappingManager::loadConfiguration( const char* filename ) {
+void KeyMappingManager::loadConfiguration( const std::string& inFilename ) {
     // if no filename is given, load default configuration
-    if ( !filename ) { filename = m_defaultConfigFile.c_str(); }
+    std::string filename = inFilename;
+    if ( filename.empty() ) { filename = m_defaultConfigFile.c_str(); }
 
     delete m_file;
-    m_file = new QFile( filename );
+    m_file = new QFile( filename.c_str() );
```
Comment (nmellado) [inline src/Gui/Utils/KeyMappingManager.cpp:236]:
```suggestion
    m_file = new QFile( QString::fromStdString( filename ) );
```

File: tests/integration/CMakeLists.txt [hunk 2]
```diff
@@ -30,3 +30,47 @@ add_test(NAME integration_${INTEGRATION_TEST}_compare_to_ref
...
+find_program(LIBXML2_XMLLINT_EXECUTABLE xmllint)
+find_program(SED_EXECUTABLE sed)
+
+if(LIBXML2_XMLLINT_EXECUTABLE AND SED_EXECUTABLE)
+    find_package(Filesystem COMPONENTS Final Experimental REQUIRED)
+    find_package(Qt5 COMPONENTS Core Widgets OpenGL Xml REQUIRED)
+
+    set(Qt5_LIBRARIES Qt5::Core Qt5::Widgets Qt5::OpenGL Qt5::Xml)
+
+    # TinyPlyFileLoader read tests
```
Comment (nmellado) [inline tests/integration/CMakeLists.txt:49]:
```suggestion
    # KeyMappingManager read tests
```

File: tests/integration/KeyMappingManager/main.cpp [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    /// need a base dir in argv[1]
+    if ( argc != 2 ) { return -1; }
+    std::string base {argv[1]};
+
+    QSettings settings;
+    settings.clear();
+    KeyMappingManager::createInstance();
+    auto mgr = Gui::KeyMappingManager::getInstance();
+    auto optionalPath {Core::Resources::getRadiumResourcesPath()};
+    auto resourcesRootDir {optionalPath.value_or( "[[Default resrouces path not found]]" )};
```
Comment (nmellado) [inline tests/integration/KeyMappingManager/main.cpp:23]:
```suggestion
    auto resourcesRootDir {optionalPath.value_or( "[[Default ressource path not found]]" )};
```

File: tests/integration/KeyMappingManager/main.cpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+using Idx = KeyMappingManager::KeyMappingAction;
+using Ctx = KeyMappingManager::Context;
+
+int main( int argc, char** argv ) {
+
+    /// need a base dir in argv[1]
+    if ( argc != 2 ) { return -1; }
+    std::string base {argv[1]};
+
+    QSettings settings;
```
Comment (nmellado) [inline tests/integration/KeyMappingManager/main.cpp:18]: Maybe set organization and application ids ? https://doc.qt.io/qt-5/qsettings.html#QSettings
Note: we have default organization in `Gui`, might would make sense to use it here for consistency.

File: tests/unittest/CMakeLists.txt [hunk 5]
```diff
@@ -21,23 +21,41 @@ add_executable(unittests
     Core/raycast.cpp
     Core/resources.cpp
     Core/string.cpp
-    Core/topomesh.cpp)
+    Core/topomesh.cpp
+    Gui/keymapping.cpp)
```
Comment (nmellado) [inline tests/unittest/CMakeLists.txt:25]:
```suggestion
    Gui/keymapping.cpp
    )
```
Very minor, just to ease next editions.

File: tests/integration/KeyMappingManager/main.cpp (hunk 4 above)
Comment (dlyr) [reply tests/integration/KeyMappingManager/main.cpp:18]: The idea here is to have "blank" settings and skip any existing one, so I'll vote for "UnitTests" organization and settings

Comment (dlyr) [reply tests/integration/KeyMappingManager/main.cpp:18]: Ok I push a proposal

File: src/Gui/Utils/KeyMappingManager.cpp [hunk 6]
```diff
@@ -29,15 +29,16 @@ KeyMappingManager::KeyMappingManager() :
 
     QSettings settings;
     QString keyMappingFilename =
-        settings.value( "keymapping/config", m_defaultConfigFile.c_str() ).toString();
-    if ( !keyMappingFilename.contains( m_defaultConfigFile.c_str() ) )
+        settings.value( "keymapping/config", QString::fromStdString( m_defaultConfigFile ) )
```
Comment (dlyr) [inline src/Gui/Utils/KeyMappingManager.cpp:32]: I have removed use of c_str here also.

File: tests/integration/KeyMappingManager/main.cpp [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+using Idx = KeyMappingManager::KeyMappingAction;
+using Ctx = KeyMappingManager::Context;
+
+int main( int argc, char** argv ) {
+
+    /// need a base dir in argv[1]
+    if ( argc != 2 ) { return -1; }
+    std::string base {argv[1]};
+
+    QSettings settings( "IntegrationTests", "KeyMappingManager" );
```
Comment (nmellado) [inline tests/integration/KeyMappingManager/main.cpp:18]:
```suggestion
    QSettings settings( "RadiumIntegrationTests", "KeyMappingManager" );
```
?

**Review by nmellado — COMMENTED** [review]
One last minor comment, but sounds good

File: tests/integration/KeyMappingManager/main.cpp (hunk 7 above)
Comment (dlyr) [reply tests/integration/KeyMappingManager/main.cpp:18]: Fine, I add it to both unit and integration tests.

**Comment by dlyr** [comment]
But I still have an untested thing, regarding my question above

**Comment by nmellado** [comment]
Ok yes, I agree, auto-saving sounds like a desirable functionality in some cases, but clearly not something we need as default.

**Comment by dlyr** [comment]
Maybe we can have this as an open issue, and discuss how it should be handled.
I'm not sure if having the KeyMappingManager asking the user directly is a good design, or if the "main application" have to check if the file is modified before loading a new one or exit.

**Comment by nmellado** [comment]
Ok, let's focus this PR on testing, and flag this by an issue as you suggest.
So we are good to merge, right ?

**Comment by dlyr** [comment]
Yes

**Review by nmellado — APPROVED** [review]

**Pull request merged**